File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
class Solution {
    
    static boolean[] vis = new boolean[51];
    static int answer = 0;
    
    public int solution(String begin, String target, String[] words) {
        
        dfs(0,begin, target, words);
        
        return answer;
    }
    
    static void dfs(int depth, String cur, String target, String[] words){
        if(cur.equals(target)){
            if(answer==0){
                answer = depth;
                return;
            }
            answer = Math.min(answer, depth);
            return;
        }
        
        for(int i=0;i<words.length;i++){
            if(vis[i]) continue;
            if(!isValid(cur, words[i])) continue;
            vis[i] = true;
            dfs(depth+1, words[i], target, words);
            vis[i] = false;
        }
    }
    
    
    static boolean isValid(String cur, String word){
        int cnt = 0;
        
        for(int i=0; i<cur.length(); i++){
            if(cur.charAt(i)!=word.charAt(i)) cnt++;
            if(cnt>=2) return false;
        }
        return true;
    }
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.kt
import kotlin.math.*

var minCnt = 100
val vis = mutableMapOf<String, Boolean>()

class Solution {
    fun solution(begin: String, target: String, words: Array<String>): Int {
        var answer = 0
        
        words.forEach{
            vis[it] = false
        }
        
        val refinedWords = words.filter{isValid(it, begin)}
        dfs(words, refinedWords, begin, target, 0)
        
        return if(minCnt==100) 0 else minCnt
    }
    
    fun dfs(words: Array<String>, refinedWords: List<String>, begin: String, target: String, cnt:Int) {
        if(begin == target) {
            minCnt = min(minCnt, cnt)
            return
        }
        
        for(i in refinedWords.indices){
            if(vis[refinedWords[i]]!!) continue
            vis[refinedWords[i]] = true
            val refinedWords2 = words.filter{isValid(it, refinedWords[i])}.filter{!vis[it]!!}
            dfs(words, refinedWords2, refinedWords[i], target, cnt+1)
            vis[refinedWords[i]] = false
        }
        
    }
    
    fun isValid(word:String, differWord:String): Boolean {
        var cnt = 0
        for(i in word.indices){
            if(word[i] != differWord[i]) cnt++
        }
        return cnt == 1
    }
}